File: bindings/julia/src/lua_runner.jl
struct LuaRunner
    ptr::Ptr{C.psr_lua_runner}
    db::Database  # Keep reference to prevent GC
end

function LuaRunner(db::Database)
    ptr = C.psr_lua_runner_new(db.ptr)
    if ptr == C_NULL
        throw(DatabaseException("Failed to create LuaRunner"))
    end
    return LuaRunner(ptr, db)
end

function run!(runner::LuaRunner, script::String)
    err = C.psr_lua_runner_run(runner.ptr, script)
    if err != C.PSR_OK
        error_ptr = C.psr_lua_runner_get_error(runner.ptr)
        if error_ptr != C_NULL
            error_msg = unsafe_string(error_ptr)
            throw(DatabaseException("Lua error: $error_msg"))
        else
            throw(DatabaseException("Lua script execution failed"))
        end
    end
    return nothing
end

function close!(runner::LuaRunner)
    C.psr_lua_runner_free(runner.ptr)
    return nothing
end

File: bindings/dart/lib/src/lua_runner.dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';

import 'ffi/bindings.dart';
import 'ffi/library_loader.dart';
import 'database.dart';
import 'exceptions.dart';

/// A Lua script runner for executing Lua scripts with access to a PSR Database.
///
/// Use [LuaRunner] to execute Lua scripts that can interact with the database
/// via the `db` global object exposed to the script.
///
/// Example:
/// ```dart
/// final db = Database.fromSchema(':memory:', 'schema.sql');
/// final lua = LuaRunner(db);
/// lua.run('''
///   db:create_element("Collection", { label = "Item 1" })
/// ''');
/// lua.dispose();
/// db.close();
/// ```
class LuaRunner {
  Pointer<psr_lua_runner_t> _ptr;
  bool _isDisposed = false;

  /// Creates a new LuaRunner for the given database.
  LuaRunner(Database db) : _ptr = nullptr {
    _ptr = bindings.psr_lua_runner_new(db.ptr);
    if (_ptr == nullptr) {
      throw const DatabaseOperationException('Failed to create LuaRunner');
    }
  }

  void _ensureNotDisposed() {
    if (_isDisposed) {
      throw const DatabaseOperationException('LuaRunner has been disposed');
    }
  }

  /// Runs a Lua script.
  ///
  /// The script has access to the database via the global `db` object.
  /// Available methods:
  /// - `db:create_element(collection, values)` - Create an element
  /// - `db:read_scalar_strings(collection, attribute)` - Read string scalars
  ///
  /// Throws [LuaException] if the script fails to execute.
  void run(String script) {
    _ensureNotDisposed();

    final arena = Arena();
    try {
      final err = bindings.psr_lua_runner_run(
        _ptr,
        script.toNativeUtf8(allocator: arena).cast(),
      );

      if (err != margaux_error_t.MARGAUX_OK) {
        final errorPtr = bindings.psr_lua_runner_get_error(_ptr);
        if (errorPtr != nullptr) {
          final errorMsg = errorPtr.cast<Utf8>().toDartString();
          throw LuaException('Lua error: $errorMsg');
        } else {
          throw const LuaException('Lua script execution failed');
        }
      }
    } finally {
      arena.releaseAll();
    }
  }

  /// Disposes the LuaRunner and frees native resources.
  void dispose() {
    if (_isDisposed) return;
    bindings.psr_lua_runner_free(_ptr);
    _isDisposed = true;
  }
}
